python/ray/external_storage.py

```diff
@@ -1,8 +1,98 @@
 import abc
 import os
+import re
+from collections import namedtuple
 from typing import List
 
 import ray
+from ray.ray_constants import DEFAULT_OBJECT_PREFIX
+
+
+class ExternalStorageURLProtocol:
+    """Class to parse/generate urls for object spilling.
+
+    There will be 2 types of URLs.
+    1. URL used internally. These urls are stored in GCS.
+    2. URL used for external storage.
+
+    This class is used as a translator layer between 1 and 2.
+    """
+
+    def __init__(self, prefix: str):
+        self._prefix = prefix
+        # NOTE: To optimize the performance, Ray fusions small
+        # objects when it stored them. object_id_stored is
+        # an object_id that is used as a filename.
+        # (job_id)-(object_id_stored)-(object_id)
+        # -(object_size_in_bytes)-(offsets_in_bytes)
+        self.internal_url_pattern = re.compile(
```

Comment:
Instead of regexes, can we use URL parameters?

```
>>> urllib.parse.urlparse("s3://foo/bar?limit=3&offset=5")
ParseResult(scheme='s3', netloc='foo', path='/bar', params='', query='limit=3&offset=5', fragment='')
>>> urllib.parse.parse_qs("limit=3&offset=5")
{'limit': ['3'], 'offset': ['5']}
```

We can strip the parameters off before passing the URL to external storage. I think this will make the URL more readable as well.

Alternatively, we can use the fragment (`s3://foo/bar#limit=3;offset=5`)

Reply:
This sounds great. I will use urllib instead